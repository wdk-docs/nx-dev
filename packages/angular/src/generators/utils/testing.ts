import type { Tree } from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import { UnitTestRunner } from '../../utils/test-runners';
import libraryGenerator from '../library/library';

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace();

  const moduleGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'module'
  );
  const componentGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );

  await libraryGenerator(tree, {
    name: libName,
    buildable: false,
    linter: Linter.EsLint,
    publishable: false,
    simpleModuleName: false,
    skipFormat: false,
    unitTestRunner: UnitTestRunner.Jest,
  });

  await componentGenerator(tree, {
    name: 'test-button',
    project: libName,
  });

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
}`
  );

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `libs/${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  await moduleGenerator(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    project: libName,
    path: `libs/${libName}/src/lib/barrel`,
    module: 'barrel',
  });

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-spread-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-anotherview',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-spread-declare/variable-spread-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableSpreadDeclareButtonComponent } from './variable-spread-declare-button/variable-spread-declare-button.component';
import { VariableSpreadDeclareViewComponent } from './variable-spread-declare-view/variable-spread-declare-view.component';
import { VariableSpreadDeclareAnotherviewComponent } from './variable-spread-declare-anotherview/variable-spread-declare-anotherview.component';

const COMPONENTS = [ 
  VariableSpreadDeclareButtonComponent, 
  VariableSpreadDeclareViewComponent 
]

@NgModule({
  imports: [CommonModule],
  declarations: [...COMPONENTS, VariableSpreadDeclareAnotherviewComponent],
})
export class VariableSpreadDeclareModule {}`
  );

  // create a module where declared components are pulled from a static member of the module
  await moduleGenerator(tree, {
    name: 'static-member-declarations',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'cmp1',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  await componentGenerator(tree, {
    name: 'cmp2',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  tree.write(
    `libs/${libName}/src/lib/static-member-declarations/static-member-declarations.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Cmp1Component } from './cmp1/cmp1.component';
import { Cmp2Component } from './cmp2/cmp2.component';

@NgModule({
  imports: [CommonModule],
  declarations: StaticMemberDeclarationsModule.COMPONENTS,
  exports: StaticMemberDeclarationsModule.COMPONENTS
})
export class StaticMemberDeclarationsModule {
  static readonly COMPONENTS = [Cmp1Component, Cmp2Component];
}`
  );

  // create another button in a nested subpath
  await moduleGenerator(tree, {
    name: 'nested',
    project: libName,
    path: `libs/${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    project: libName,
    module: 'nested',
    path: `libs/${libName}/src/lib/nested`,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    project: libName,
  });

  return tree;
}
