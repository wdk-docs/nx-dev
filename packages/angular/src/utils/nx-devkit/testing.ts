import type { Tree } from '@nrwl/devkit';
import { names, updateJson } from '@nrwl/devkit';

export interface AppConfig {
  appName: string; // name of app
  appModule: string; // app/app.module.ts in the above sourceDir
}
export interface LibConfig {
  name: string;
  module: string;
  barrel: string;
}

var appConfig: AppConfig; // configure built in createApp()
var libConfig: LibConfig;

export function getAppConfig(): AppConfig {
  return appConfig;
}
export function getLibConfig(): LibConfig {
  return libConfig;
}

export function createApp(
  tree: Tree,
  appName: string,
  routing: boolean = true
) {
  appName = names(appName).fileName;
  // save for getAppDir() lookup by external *.spec.ts tests
  appConfig = {
    appName,
    appModule: `/apps/${appName}/src/app/app.module.ts`,
  };

  tree.write(
    appConfig.appModule,
    `
     import { NgModule } from '@angular/core';
     import { BrowserModule } from '@angular/platform-browser';
     ${routing ? "import { RouterModule } from '@angular/router'" : ''};
     import { AppComponent } from './app.component';
     @NgModule({
       imports: [BrowserModule, ${routing ? 'RouterModule.forRoot([])' : ''}],
       declarations: [AppComponent],
       bootstrap: [AppComponent]
     })
     export class AppModule {}
  `
  );
  tree.write(
    `/apps/${appName}/src/main.ts`,
    `
    import { enableProdMode } from '@angular/core';
    import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';

    import { AppModule } from './app/app.module';
    import { environment } from './environments/environment';

    if (environment.production) {
      enableProdMode();
    }

    platformBrowserDynamic()
      .bootstrapModule(AppModule)
      .catch(err => console.log(err));
  `
  );
  tree.write(
    `/apps/${appName}/tsconfig.app.json`,
    JSON.stringify({
      include: ['**/*.ts'],
    })
  );
  tree.write(
    `/apps/${appName}-e2e/tsconfig.e2e.json`,
    JSON.stringify({
      include: ['../**/*.ts'],
    })
  );
  tree.write(
    '/workspace.json',
    JSON.stringify({
      newProjectRoot: '',
      version: 1,
      projects: {
        [appName]: {
          root: `apps/${appName}`,
          sourceRoot: `apps/${appName}/src`,
          architect: {
            build: {
              options: {
                main: `apps/${appName}/src/main.ts`,
              },
            },
            serve: {
              options: {},
            },
          },
          tags: [],
        },
      },
    })
  );
}

export function createLib(tree: Tree, libName: string) {
  const { name, className, fileName, propertyName } = names(libName);

  libConfig = {
    name,
    module: `/libs/${propertyName}/src/lib/${fileName}.module.ts`,
    barrel: `/libs/${propertyName}/src/index.ts`,
  };

  tree.write(
    libConfig.module,
    `
      import { NgModule } from '@angular/core';
      import { CommonModule } from '@angular/common';
      @NgModule({
        imports: [
          CommonModule
        ],
        providers: []
      })
      export class ${className}Module { }
  `
  );
  tree.write(
    libConfig.barrel,
    `
    export * from './lib/${fileName}.module';
  `
  );
}
