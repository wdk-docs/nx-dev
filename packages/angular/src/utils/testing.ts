import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/angular',
  join(__dirname, '../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/workspace',
  join(__dirname, '../../../workspace/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/storybook',
  join(__dirname, '../../../storybook/generators.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../migrations.json')
);

export function runMigration<SchemaOptions = any>(
  schematicName: string,
  options: SchemaOptions,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync(schematicName, options, tree)
    .toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}
