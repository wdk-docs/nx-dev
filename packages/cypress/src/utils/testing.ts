import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Tree, Rule } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/cypress',
  join(__dirname, '../../generators.json')
);

const migrationRunner = new SchematicTestRunner(
  '@nrwl/cypress/migrations',
  join(__dirname, '../../migrations.json')
);

export function runSchematic(schematicName: string, options: any, tree: Tree) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(migrationName: string, options: any, tree: Tree) {
  return migrationRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}
