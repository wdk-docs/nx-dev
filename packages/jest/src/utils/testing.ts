import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Tree, Rule } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/jest',
  join(__dirname, '../../generators.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/jest/migrations',
  join(__dirname, '../../migrations.json')
);

export function runSchematic<T = any>(
  schematicName: string,
  options: T,
  tree: Tree
) {
  return testRunner
    .runSchematicAsync<T>(schematicName, options, tree)
    .toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(
  migrationName: string,
  options: unknown,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}
