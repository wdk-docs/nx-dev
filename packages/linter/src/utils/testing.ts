import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { join } from 'path';
import { Tree } from '@angular-devkit/schematics';

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/linter-migrations',
  join(__dirname, '../../migrations.json')
);

export function runMigration<T = any>(
  migrationName: string,
  options: T,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync<T>(migrationName, options, tree)
    .toPromise();
}
