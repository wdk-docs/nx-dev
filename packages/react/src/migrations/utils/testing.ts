import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';
import { updateWorkspace } from '@nrwl/workspace/src/utils/workspace';
import { readJsonInTree } from '@nrwl/workspace';
import { names } from '@nrwl/devkit';

const testRunner = new SchematicTestRunner(
  '@nrwl/react',
  join(__dirname, '../../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/storybook',
  join(__dirname, '../../../../storybook/generators.json')
);

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function updateNxJson(tree, update: (json: any) => any) {
  const updated = update(readJsonInTree(tree, '/nx.json'));
  tree.overwrite('/nx.json', JSON.stringify(updated));
}

export function createApp(tree: Tree, appName: string): Promise<Tree> {
  const { fileName } = names(appName);

  tree.create(
    `/apps/${fileName}/src/main.tsx`,
    `import ReactDOM from 'react-dom';\n`
  );

  updateNxJson(tree, (json) => {
    json.projects[appName] = { tags: [] };
    return json;
  });

  return callRule(
    updateWorkspace((workspace) => {
      workspace.projects.add({
        name: fileName,
        root: `apps/${fileName}`,
        projectType: 'application',
        sourceRoot: `apps/${fileName}/src`,
        targets: {},
      });
    }),
    tree
  );
}

export function createWebApp(tree: Tree, appName: string): Promise<Tree> {
  const { fileName } = names(appName);

  tree.create(`/apps/${fileName}/src/index.ts`, `\n`);

  updateNxJson(tree, (json) => {
    json.projects[appName] = { tags: [] };
    return json;
  });

  return callRule(
    updateWorkspace((workspace) => {
      workspace.projects.add({
        name: fileName,
        root: `apps/${fileName}`,
        projectType: 'application',
        sourceRoot: `apps/${fileName}/src`,
        targets: {},
      });
    }),
    tree
  );
}

export function createLib(tree: Tree, libName: string): Promise<Tree> {
  const { fileName } = names(libName);

  tree.create(`/libs/${fileName}/src/index.ts`, `import React from 'react';\n`);

  updateNxJson(tree, (json) => {
    json.projects[libName] = { tags: [] };
    return json;
  });

  return callRule(
    updateWorkspace((workspace) => {
      workspace.projects.add({
        name: fileName,
        root: `libs/${fileName}`,
        projectType: 'library',
        sourceRoot: `libs/${fileName}/src`,
        targets: {},
      });
    }),
    tree
  );
}
