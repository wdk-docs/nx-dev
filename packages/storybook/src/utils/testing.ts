import { join } from 'path';
import { externalSchematic, Rule, Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';

import { createEmptyWorkspace } from '@nrwl/workspace/testing';

const testRunner = new SchematicTestRunner(
  '@nrwl/storybook',
  join(__dirname, '../../generators.json')
);

['angular', 'react', 'jest', 'cypress'].forEach((collection) =>
  testRunner.registerCollection(
    `@nrwl/${collection}`,
    join(__dirname, `../../../${collection}/generators.json`)
  )
);

const migrationRunner = new SchematicTestRunner(
  '@nrwl/storybook/migrations',
  join(__dirname, '../../migrations.json')
);

export function runSchematic(schematicName: string, options: any, tree: Tree) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function runExternalSchematic(
  collectionName: string,
  schematicName: string,
  options: any,
  tree: Tree
) {
  return testRunner
    .runExternalSchematicAsync(collectionName, schematicName, options, tree)
    .toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(migrationName: string, options: any, tree: Tree) {
  return migrationRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}

export async function createTestUILib(
  libName: string,
  collectionName: '@nrwl/angular' | '@nrwl/react',
  options: any = {}
): Promise<Tree> {
  let appTree = Tree.empty();
  appTree = createEmptyWorkspace(appTree);
  appTree = await callRule(
    externalSchematic(collectionName, 'library', {
      name: libName,
      ...options,
    }),
    appTree
  );
  appTree = await callRule(
    externalSchematic(collectionName, 'component', {
      name: 'test-button',
      project: libName,
    }),
    appTree
  );

  if (collectionName === '@nrwl/angular') {
    updateAngularComponent(appTree);
  }
  if (collectionName === '@nrwl/react') {
    // @TODO
  }

  appTree = await callRule(
    externalSchematic(collectionName, 'component', {
      name: 'test-other',
      project: libName,
    }),
    appTree
  );

  return appTree;

  function updateAngularComponent(appTree: Tree) {
    appTree.overwrite(
      `libs/${libName}/src/lib/test-button/test-button.component.ts`,
      `
import { Component, OnInit, Input } from '@angular/core';
import { tmpdir } from 'os';
import { mkdtempSync } from 'fs';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent implements OnInit {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age: number;
  @Input() isOn = false;

  constructor() { }

  ngOnInit() {
  }

}
`
    );
    appTree.overwrite(
      `libs/${libName}/src/lib/test-button/test-button.component.html`,
      `<button [attr.type]="type" [ngClass]="style"></button>`
    );
  }
}
