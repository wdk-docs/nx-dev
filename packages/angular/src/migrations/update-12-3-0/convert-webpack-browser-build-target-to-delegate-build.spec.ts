import { addProjectConfiguration, readJson, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import convertWebpackBrowserBuildTargetToDelegateBuild from './convert-webpack-browser-build-target-to-delegate-build';

function getWsConfig(tree: Tree) {
  return readJson(tree, 'workspace.json');
}

describe('convertWebpackBrowserBuildTargetToDelegateBuild', () => {
  let tree: Tree;

  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace();

    addProjectConfiguration(tree, 'ng-app1', {
      root: '',
      sourceRoot: 'src',
      projectType: 'application',
      targets: {
        customBuild: {
          executor: '@angular-devkit/build-angular:browser',
        },
        build: {
          executor: '@nrwl/angular:webpack-browser',
          options: {
            buildTarget: 'ng-app:customBuild',
            outputPath: 'dist/apps/ng-app1',
            tsConfig: 'apps/ng-app1/tsconfig.app.json',
            index: 'apps/ng-app1/src/index.html',
            main: 'apps/ng-app1/src/main.ts',
            polyfills: 'apps/ng-app1/src/polyfills.ts',
            inlineStyleLanguage: 'scss',
            assets: ['apps/ng-app1/src/favicon.ico', 'apps/ng-app1/src/assets'],
            styles: ['apps/ng-app1/src/styles.scss'],
            scripts: [],
          },
          configurations: {
            production: {
              budgets: [
                {
                  type: 'initial',
                  maximumWarning: '500kb',
                  maximumError: '1mb',
                },
                {
                  type: 'anyComponentStyle',
                  maximumWarning: '2kb',
                  maximumError: '4kb',
                },
              ],
              fileReplacements: [
                {
                  replace: 'apps/ng-app1/src/environments/environment.ts',
                  with: 'apps/ng-app1/src/environments/environment.prod.ts',
                },
              ],
              outputHashing: 'all',
            },
            development: {
              buildOptimizer: false,
              optimization: false,
              vendorChunk: true,
              extractLicenses: false,
              sourceMap: true,
              namedChunks: true,
            },
          },
          defaultConfiguration: 'production',
        },
      },
    });

    addProjectConfiguration(tree, 'ng-app2', {
      root: '',
      sourceRoot: 'src',
      projectType: 'application',
      targets: {
        customBuild: {
          executor: '@angular-devkit/build-angular:browser',
          options: {
            outputPath: 'dist/apps/ng-app2',
            tsConfig: 'apps/ng-app2/tsconfig.app.json',
            index: 'apps/ng-app2/src/index.html',
            main: 'apps/ng-app2/src/different-main.ts',
            additionalProperty: 'foo',
          },
          configurations: {
            development: {
              buildOptimizer: true,
            },
            production: {
              additionalProperty: 'bar',
              fileReplacements: [
                {
                  replace: 'apps/ng-app2/src/environments/environment.ts',
                  with: 'apps/ng-app2/src/environments/environment.other.ts',
                },
              ],
            },
          },
        },
        build: {
          executor: '@nrwl/angular:webpack-browser',
          options: {
            outputPath: 'dist/apps/ng-app2',
            tsConfig: 'apps/ng-app2/tsconfig.app.json',
            index: 'apps/ng-app2/src/index.html',
            main: 'apps/ng-app2/src/main.ts',
            polyfills: 'apps/ng-app2/src/polyfills.ts',
            inlineStyleLanguage: 'scss',
            assets: ['apps/ng-app2/src/favicon.ico', 'apps/ng-app2/src/assets'],
            styles: ['apps/ng-app2/src/styles.scss'],
            scripts: [],
          },
          configurations: {
            production: {
              buildTarget: 'ng-app:customBuild:production',
              budgets: [
                {
                  type: 'initial',
                  maximumWarning: '500kb',
                  maximumError: '1mb',
                },
                {
                  type: 'anyComponentStyle',
                  maximumWarning: '2kb',
                  maximumError: '4kb',
                },
              ],
              fileReplacements: [
                {
                  replace: 'apps/ng-app2/src/environments/environment.ts',
                  with: 'apps/ng-app2/src/environments/environment.prod.ts',
                },
              ],
              outputHashing: 'all',
            },
            development: {
              buildTarget: 'ng-app:customBuild:development',
              buildOptimizer: false,
              optimization: false,
              vendorChunk: true,
              extractLicenses: false,
              sourceMap: true,
              namedChunks: true,
            },
          },
          defaultConfiguration: 'production',
        },
      },
    });

    addProjectConfiguration(tree, 'ng-app3', {
      root: '',
      sourceRoot: 'src',
      projectType: 'application',
      targets: {
        build: {
          executor: '@nrwl/angular:webpack-browser',
          options: {
            outputPath: 'dist/apps/ng-app3',
            tsConfig: 'apps/ng-app3/tsconfig.app.json',
            index: 'apps/ng-app3/src/index.html',
            main: 'apps/ng-app3/src/main.ts',
            polyfills: 'apps/ng-app3/src/polyfills.ts',
            inlineStyleLanguage: 'scss',
            assets: ['apps/ng-app3/src/favicon.ico', 'apps/ng-app3/src/assets'],
            styles: ['apps/ng-app3/src/styles.scss'],
            scripts: [],
          },
          configurations: {
            production: {
              budgets: [
                {
                  type: 'initial',
                  maximumWarning: '500kb',
                  maximumError: '1mb',
                },
                {
                  type: 'anyComponentStyle',
                  maximumWarning: '2kb',
                  maximumError: '4kb',
                },
              ],
              fileReplacements: [
                {
                  replace: 'apps/ng-app3/src/environments/environment.ts',
                  with: 'apps/ng-app3/src/environments/environment.prod.ts',
                },
              ],
              outputHashing: 'all',
            },
            development: {
              buildOptimizer: false,
              optimization: false,
              vendorChunk: true,
              extractLicenses: false,
              sourceMap: true,
              namedChunks: true,
            },
          },
          defaultConfiguration: 'production',
        },
      },
    });
  });

  it('should update the configuration correctly', async () => {
    await convertWebpackBrowserBuildTargetToDelegateBuild(tree);

    expect(getWsConfig(tree)).toMatchSnapshot();
  });
});
