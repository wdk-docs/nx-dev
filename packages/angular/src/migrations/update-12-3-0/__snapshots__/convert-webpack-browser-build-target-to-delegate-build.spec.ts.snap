// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertWebpackBrowserBuildTargetToDelegateBuild should update the configuration correctly 1`] = `
Object {
  "projects": Object {
    "ng-app1": Object {
      "architect": Object {
        "build": Object {
          "builder": "@nrwl/angular:delegate-build",
          "configurations": Object {
            "development": Object {
              "buildTarget": "ng-app1:customBuild:development",
            },
            "production": Object {
              "buildTarget": "ng-app1:customBuild:production",
            },
          },
          "defaultConfiguration": "production",
          "options": Object {
            "buildTarget": "ng-app:customBuild",
            "outputPath": "dist/apps/ng-app1",
            "tsConfig": "apps/ng-app1/tsconfig.app.json",
          },
        },
        "customBuild": Object {
          "builder": "@angular-devkit/build-angular:browser",
          "configurations": Object {
            "development": Object {
              "buildOptimizer": false,
              "extractLicenses": false,
              "namedChunks": true,
              "optimization": false,
              "sourceMap": true,
              "vendorChunk": true,
            },
            "production": Object {
              "budgets": Array [
                Object {
                  "maximumError": "1mb",
                  "maximumWarning": "500kb",
                  "type": "initial",
                },
                Object {
                  "maximumError": "4kb",
                  "maximumWarning": "2kb",
                  "type": "anyComponentStyle",
                },
              ],
              "fileReplacements": Array [
                Object {
                  "replace": "apps/ng-app1/src/environments/environment.ts",
                  "with": "apps/ng-app1/src/environments/environment.prod.ts",
                },
              ],
              "outputHashing": "all",
            },
          },
          "options": Object {
            "assets": Array [
              "apps/ng-app1/src/favicon.ico",
              "apps/ng-app1/src/assets",
            ],
            "index": "apps/ng-app1/src/index.html",
            "inlineStyleLanguage": "scss",
            "main": "apps/ng-app1/src/main.ts",
            "polyfills": "apps/ng-app1/src/polyfills.ts",
            "scripts": Array [],
            "styles": Array [
              "apps/ng-app1/src/styles.scss",
            ],
          },
        },
      },
      "projectType": "application",
      "root": "",
      "sourceRoot": "src",
    },
    "ng-app2": Object {
      "architect": Object {
        "build": Object {
          "builder": "@nrwl/angular:delegate-build",
          "configurations": Object {
            "development": Object {
              "buildTarget": "ng-app:customBuild:development",
            },
            "production": Object {
              "buildTarget": "ng-app:customBuild:production",
            },
          },
          "defaultConfiguration": "production",
          "options": Object {
            "outputPath": "dist/apps/ng-app2",
            "tsConfig": "apps/ng-app2/tsconfig.app.json",
          },
        },
        "customBuild": Object {
          "builder": "@angular-devkit/build-angular:browser",
          "configurations": Object {
            "development": Object {
              "buildOptimizer": false,
              "extractLicenses": false,
              "namedChunks": true,
              "optimization": false,
              "sourceMap": true,
              "vendorChunk": true,
            },
            "production": Object {
              "additionalProperty": "bar",
              "budgets": Array [
                Object {
                  "maximumError": "1mb",
                  "maximumWarning": "500kb",
                  "type": "initial",
                },
                Object {
                  "maximumError": "4kb",
                  "maximumWarning": "2kb",
                  "type": "anyComponentStyle",
                },
              ],
              "fileReplacements": Array [
                Object {
                  "replace": "apps/ng-app2/src/environments/environment.ts",
                  "with": "apps/ng-app2/src/environments/environment.prod.ts",
                },
              ],
              "outputHashing": "all",
            },
          },
          "options": Object {
            "additionalProperty": "foo",
            "assets": Array [
              "apps/ng-app2/src/favicon.ico",
              "apps/ng-app2/src/assets",
            ],
            "index": "apps/ng-app2/src/index.html",
            "inlineStyleLanguage": "scss",
            "main": "apps/ng-app2/src/main.ts",
            "polyfills": "apps/ng-app2/src/polyfills.ts",
            "scripts": Array [],
            "styles": Array [
              "apps/ng-app2/src/styles.scss",
            ],
          },
        },
      },
      "projectType": "application",
      "root": "",
      "sourceRoot": "src",
    },
    "ng-app3": Object {
      "architect": Object {
        "build": Object {
          "builder": "@nrwl/angular:webpack-browser",
          "configurations": Object {
            "development": Object {
              "buildOptimizer": false,
              "extractLicenses": false,
              "namedChunks": true,
              "optimization": false,
              "sourceMap": true,
              "vendorChunk": true,
            },
            "production": Object {
              "budgets": Array [
                Object {
                  "maximumError": "1mb",
                  "maximumWarning": "500kb",
                  "type": "initial",
                },
                Object {
                  "maximumError": "4kb",
                  "maximumWarning": "2kb",
                  "type": "anyComponentStyle",
                },
              ],
              "fileReplacements": Array [
                Object {
                  "replace": "apps/ng-app3/src/environments/environment.ts",
                  "with": "apps/ng-app3/src/environments/environment.prod.ts",
                },
              ],
              "outputHashing": "all",
            },
          },
          "defaultConfiguration": "production",
          "options": Object {
            "assets": Array [
              "apps/ng-app3/src/favicon.ico",
              "apps/ng-app3/src/assets",
            ],
            "index": "apps/ng-app3/src/index.html",
            "inlineStyleLanguage": "scss",
            "main": "apps/ng-app3/src/main.ts",
            "outputPath": "dist/apps/ng-app3",
            "polyfills": "apps/ng-app3/src/polyfills.ts",
            "scripts": Array [],
            "styles": Array [
              "apps/ng-app3/src/styles.scss",
            ],
            "tsConfig": "apps/ng-app3/tsconfig.app.json",
          },
        },
      },
      "projectType": "application",
      "root": "",
      "sourceRoot": "src",
    },
  },
  "version": 1,
}
`;
